db/commitlog/commitlog.cc

```diff
@@ -2702,14 +2779,112 @@ db::commitlog::read_log_file(sstring filename, sstring pfx, commit_load_reader_f
+            auto rem = buffer.size_bytes();
+            auto vec = std::move(buffer).release();
+            auto block = align_up(pos - initial.size_bytes(), alignment);
+
+            while (rem < size) {
+                if (eof) {
+                    throw segment_truncation(block);
+                }
+
+                auto block_size = alignment - initial.size_bytes();
+                // using a stream is perhaps not 100% effective, but we need to 
+                // potentially address data in pages smaller than the current 
+                // disk/fs we are reading from can handle (but please no). 
+                auto tmp = co_await frag_reader.read_exactly(fin, block_size);
+
+                if (tmp.size_bytes() == 0) {
+                    eof = true;
+                    throw segment_truncation(block);
+                }
+
+                crc32_nbo crc;
+                // crc all but the final crc
+                size_t n = block_size - sizeof(uint32_t);
+
+                bool all_zero = true;
+
+                if (!initial.empty()) {
+                    for (auto& bv : initial) {
+                        crc.process_bytes(bv.get(), bv.size());
+                    }
+                    initial = {};
+                    all_zero = false;
+                }
+
+                for (auto& bv : tmp) {
+                    auto np = std::min(bv.size(), n);
+                    crc.process_bytes(bv.get(), np);
+                    all_zero &= std::all_of(bv.get(), bv.get() + bv.size(), [](char c) { return c == 0; });
+                    n -= np;
+                }
+
+                block += alignment;
+
+                if (!all_zero) {
+                    auto in = tmp.get_istream();
+                    in.skip(block_size - detail::sector_overhead_size);
+
+                    auto id = read<uint64_t>(in);
+                    auto check = read<uint32_t>(in);
+                    auto checksum = crc.checksum();
+
+                    if (check != checksum) {
+                        throw segment_data_corruption_error("Data corruption", alignment);
+                    }
+                    if (id != this->id) {
+                        throw segment_truncation(pos + rem);
+                    }
+                }
+                tmp.remove_suffix(detail::sector_overhead_size);
+
+                rem += tmp.size_bytes();
+                pos += detail::sector_overhead_size;
+
+                auto vec2 = std::move(tmp).release();
+                for (auto&& v : vec2) {
+                    vec.emplace_back(std::move(v));
+                }
             }
 
+            decltype(vec) next;
+
+            auto leave = rem - size;
+            size_t left = 0;
+            while (left < leave) {
+                auto tmp = std::move(vec.back());
+                vec.pop_back();
+                auto s = tmp.size();
+                if (s <= leave) {
+                    next.emplace_back(std::move(tmp));
```

Comment:
Aren't we filling the `next` vector in reverse here? Or is it expected to be reversed